File: python/automarketing/video/aivideo.py
"""
AI video clip — Google nu Gemini Omni Flash (Google Flow ma "Omni").

Aa file ffmpeg ni JAGYA E nathi. Reel to ffmpeg j banave che — e sasti,
zadapi ane bharoso layak rit che. Pan ketlak shot ma KHAREKHAR halchal
joiye che: kapdu udtu hoy, model chaale, product ferve. Ena mate aa file.

Rasto jaani joine "image → video" j rakhyo che:

    product ni photo → gpt-image-1 (scene ni image) → Omni (e j image halave)

Sidha text thi video banaviye to AI product ne potani rite kalpi le che ane
kapdu/rang badlai jaay che. Pehla image banavi ne pachi ENE J halavvathi
product jem no tem rahe che — vechan mate aa j ek maatra saacho rasto che.

API — Interactions API (https://ai.google.dev/gemini-api/docs/omni):
    POST /v1beta/interactions      → video (3-10s, 720p, 24fps)
    delivery "uri" → files/<id>    → ACTIVE thay tya sudhi rah → download
"""

from __future__ import annotations

import asyncio
import base64
import re
import time
from dataclasses import dataclass
from typing import Any, Literal, Optional

from ..config import settings
from ..errors import FatalError, NotConfigured, RetryableError
from ..media import images as imagelib
from ..pipeline.chain import Candidate, ChainResult, run_chain
from ..pipeline.http import get_client, request_bytes

GENAI_BASE = "https://generativelanguage.googleapis.com/v1beta"

#: Omni ni potani had — aa thi bahar magiye to e potani rite kaapi de che.
MIN_CLIP_SECONDS = 3
MAX_CLIP_SECONDS = 10

VideoAspectRatio = Literal["9:16", "16:9"]

OMNI_HINT = (
    "GEMINI_API_KEY joiye ane Omni tamara Google account ma chalu hovu joiye "
    "(paid tier). Aa vagar pan reel banse — ffmpeg still image ne halave che."
)


@dataclass
class GeneratedClip:
    data: bytes
    mime_type: str
    provider: str
    prompt: str


# ------------------------------------------------------------------ #
#  Config
# ------------------------------------------------------------------ #


def ai_video_configured() -> bool:
    """AI video chalu che ke nahi. `AI_VIDEO_ENABLED=false` thi bandh."""
    return settings.ai_video_enabled


def ai_video_max_clips() -> int:
    """Ek reel ma ketla AI clip banavva (0 = sav bandh)."""
    return settings.ai_video_max_clips


# ------------------------------------------------------------------ #
#  Jawab mathi video kaadhvu
# ------------------------------------------------------------------ #


def _find_video(payload: dict) -> Optional[dict]:
    """Video kya paḍyo che e shodhe — inline base64 ke Files API no URI."""
    output = payload.get("output_video") or {}
    if output.get("uri"):
        return {"uri": output["uri"], "mime_type": output.get("mime_type")}

    for step in payload.get("steps") or []:
        for content in step.get("content") or []:
            if not isinstance(content, dict):
                continue
            if content.get("type") == "video" and (content.get("data") or content.get("uri")):
                return content

    return None


def _file_id(uri: str) -> str:
    """
    Files API nu id. URI gme te aakar no hoy shake:
        files/abc123
        https://…/v1beta/files/abc123:download?alt=media
    """
    match = re.search(r"files/([^:/?#]+)", uri or "")
    return match.group(1) if match else ""


async def _wait_for_file(file_id: str) -> None:
    """
    File taiyar (ACTIVE) thay tya sudhi rah jue che.

    Omni motu video Files API ma mukе che ane e turant download layak hotu
    nathi — sidhu download karie to khali ke adhuri file aave.
    """
    client = get_client()
    deadline = time.monotonic() + 300.0

    while True:
        response = await client.get(
            f"{GENAI_BASE}/files/{file_id}",
            params={"key": settings.gemini_key},
            timeout=30.0,
        )
        if response.status_code >= 400:
            raise RetryableError(f"Omni file check: HTTP {response.status_code}")

        payload = response.json() or {}
        state = payload.get("state")

        if state in (None, "", "ACTIVE"):
            return
        if state == "FAILED":
            error = (payload.get("error") or {}).get("message") or "FAILED"
            raise FatalError(f"Omni: file process na thai — {error}")
        if time.monotonic() > deadline:
            raise RetryableError("Omni: video taiyar thata bahu var lagi")

        await asyncio.sleep(3.0)


# ------------------------------------------------------------------ #
#  Omni
# ------------------------------------------------------------------ #


async def _omni_clip(
    prompt: str,
    image: Optional[bytes],
    aspect: VideoAspectRatio,
    duration: int,
) -> GeneratedClip:
    api_key = settings.gemini_key
    if not api_key:
        raise NotConfigured(OMNI_HINT)

    model = settings.omni_model

    # Omni pase `duration` no parameter nathi — e prompt mathi j samje che.
    prompt_text = "\n".join(
        [
            prompt.strip(),
            "",
            f"Duration: about {duration} seconds. One single continuous shot, no cuts.",
            "Keep every product, garment, colour, print and face exactly as in the "
            "reference image — animate it, do not redesign it.",
            "No on-screen text, no captions, no subtitles, no logos, no watermarks.",
        ]
    )

    parts: list[dict[str, Any]] = []
    if image:
        b64, _ = imagelib.prepare_for_vision(image, max_side=1024)
        parts.append({"type": "image", "mime_type": "image/jpeg", "data": b64})
    parts.append({"type": "text", "text": prompt_text})

    client = get_client()
    try:
        response = await client.post(
            f"{GENAI_BASE}/interactions",
            params={"key": api_key},
            json={
                "model": model,
                # Fakt text hoy to Omni sadhi string pan sweekare che, pan
                # array hamesha chale che — etle ek j rasto rakhie chie.
                "input": parts,
                "response_format": {
                    "type": "video",
                    "aspect_ratio": aspect,
                    # Video mota bhage 4MB thi motu hoy che ane tyare inline
                    # base64 aavtu nathi — etle hamesha URI j magie chie.
                    "delivery": "uri",
                },
            },
            timeout=360.0,
        )
    except Exception as error:  # noqa: BLE001
        raise RetryableError(f"Omni sudhi pahonchi na shakaya: {error}") from error

    if response.status_code >= 400:
        message = f"Omni: HTTP {response.status_code} {response.text[:300]}"
        if response.status_code == 429 or response.status_code >= 500:
            raise RetryableError(message)
        raise FatalError(message)

    payload = response.json() or {}

    if payload.get("error"):
        error = payload["error"]
        raise FatalError(f"Omni: {error.get('message') or error.get('status') or error}")

    status = payload.get("status")
    if status and status != "completed":
        raise RetryableError(f"Omni: status {status}")

    video = _find_video(payload)
    if not video:
        raise RetryableError("Omni e video na aapyu")

    mime_type = video.get("mime_type") or "video/mp4"

    if video.get("data"):
        data = base64.b64decode(video["data"])
    else:
        file_id = _file_id(video.get("uri") or "")
        if not file_id:
            raise RetryableError(f"Omni no video URI samjayo nahi: {video.get('uri')}")

        await _wait_for_file(file_id)

        data = await request_bytes(
            f"{GENAI_BASE}/files/{file_id}:download",
            params={"alt": "media", "key": api_key},
            timeout=300.0,
        )

    if len(data) < 10_000:
        raise RetryableError("Omni e khali file aapi")

    return GeneratedClip(
        data=data,
        mime_type=mime_type,
        provider=f"omni:{model}",
        prompt=prompt,
    )


# ------------------------------------------------------------------ #
#  Public API
# ------------------------------------------------------------------ #


async def generate_video_clip(
    prompt: str,
    *,
    image: Optional[bytes] = None,
    aspect: VideoAspectRatio = "9:16",
    duration: float = 5.0,
    prefer: str = "",
) -> ChainResult[GeneratedClip]:
    """
    Ek AI video clip banave.

    Fail thay to `ChainError` phenke che — call karnaar e pakadi ne still
    image par pachho vali javu joiye. Reel aana karane kyarey atakvu na joiye.
    """
    seconds = int(round(max(MIN_CLIP_SECONDS, min(MAX_CLIP_SECONDS, duration))))

    return await run_chain(
        [
            Candidate[GeneratedClip](
                name="omni",
                label="Gemini Omni Flash",
                free=False,
                configured=ai_video_configured,
                run=lambda: _omni_clip(prompt, image, aspect, seconds),
                # 10s no clip banta ~1-2 minute lage che.
                timeout=380.0,
            )
        ],
        label="AI video",
        prefer=(prefer or settings.preferred_video_provider),
        prefer_free=False,
        retries=1,
        backoff=5.0,
    )


def aivideo_status() -> list[dict]:
    """Setup page mate — AI video taiyar che ke nahi."""
    ready = ai_video_configured()
    return [
        {
            "key": "omni",
            "label": f"Gemini Omni Flash ({settings.omni_model})",
            "free": False,
            "configured": ready,
            "note": (
                f"Reel dith {ai_video_max_clips()} clip sudhi — AI_VIDEO_MAX_CLIPS thi badlo"
                if ready
                else OMNI_HINT
            ),
        }
    ]

File: src/lib/video/ai-video.ts
/**
 * AI video clips — Omni (Google's Gemini Omni Flash).
 *
 * This does NOT replace ffmpeg. ffmpeg builds the reel by moving still images
 * (a Ken Burns pan), which is cheap, fast and completely reliable. But a few
 * shots genuinely need real motion: fabric catching the air, a model turning,
 * a product actually being used. That is what this file is for.
 *
 * The route is deliberately image-to-video, never text-to-video:
 *
 *     product photo -> a scene image -> Omni animates that exact image
 *
 * Generate straight from text and the model invents its own version of the
 * product; the fabric changes, the colour drifts, the print is wrong. Building
 * the still first and animating *that* keeps the product identical, which is
 * the only acceptable outcome when the video exists to sell the thing.
 *
 * API: the Interactions API — https://ai.google.dev/gemini-api/docs/omni
 *   POST /v1beta/interactions   -> a 3-10 second clip, 720p, 24fps
 *   delivery:"uri" -> files/<id> -> poll until ACTIVE -> download
 */

import sharp from "sharp";

import {
  runChain,
  apiFetch,
  FatalError,
  RetryableError,
  type ChainResult,
} from "@/lib/pipeline/chain";

const GENAI_BASE = "https://generativelanguage.googleapis.com/v1beta";

/** Omni's own limits — ask for anything outside this and it clamps silently. */
export const MIN_CLIP_SECONDS = 3;
export const MAX_CLIP_SECONDS = 10;

export type VideoAspectRatio = "9:16" | "16:9";

export type GeneratedClip = {
  data: Buffer;
  mimeType: string;
  provider: string;
  prompt: string;
};

export type GenerateClipOptions = {
  /** What should be visible, and how the camera moves. */
  prompt: string;
  /** The image to animate. Supplying it is what keeps the product unchanged. */
  image?: { data: Buffer; mimeType: string };
  aspectRatio?: VideoAspectRatio;
  /** Between 3 and 10 seconds. Omni takes this from the prompt text. */
  durationSeconds?: number;
  prefer?: string;
};

/* ------------------------------------------------------------------ *
 *  Configuration
 * ------------------------------------------------------------------ */

function geminiKey(): string {
  return process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY || "";
}

function omniModel(): string {
  return process.env.OMNI_MODEL || "gemini-omni-flash-preview";
}

/** Whether AI video is switched on. Default: on whenever a key exists. */
export function aiVideoConfigured(): boolean {
  if (process.env.AI_VIDEO_ENABLED === "false") return false;
  return Boolean(geminiKey());
}

/**
 * How many AI clips to put in one reel.
 *
 * Each clip takes a minute or two and costs money, so animating every scene
 * makes little sense. Two or three clips on the shots that matter most — the
 * hook and the lifestyle beat — with stills for the rest gives the best result
 * per rupee. Set it to 0 to turn AI video off entirely.
 */
export function aiVideoMaxClips(): number {
  const raw = Number(process.env.AI_VIDEO_MAX_CLIPS);
  if (!Number.isFinite(raw) || raw < 0) return 2;
  return Math.min(8, Math.floor(raw));
}

/* ------------------------------------------------------------------ *
 *  Omni — the Interactions API
 * ------------------------------------------------------------------ */

type InteractionContent = {
  type?: string;
  mime_type?: string;
  data?: string;
  uri?: string;
  text?: string;
};

type InteractionResponse = {
  id?: string;
  status?: string;
  steps?: Array<{ type?: string; content?: InteractionContent[] }>;
  output_video?: { uri?: string; mime_type?: string };
  error?: { message?: string; status?: string };
};

/** Finds the video in the response — inline base64 or a Files API URI. */
function findVideo(json: InteractionResponse): InteractionContent | null {
  if (json.output_video?.uri) {
    return {
      type: "video",
      uri: json.output_video.uri,
      mime_type: json.output_video.mime_type,
    };
  }
  for (const step of json.steps ?? []) {
    for (const content of step.content ?? []) {
      if (content.type === "video" && (content.data || content.uri)) return content;
    }
  }
  return null;
}

/**
 * The Files API id. The URI arrives in more than one shape:
 *   files/abc123
 *   https://…/v1beta/files/abc123:download?alt=media
 */
function fileIdFromUri(uri: string): string | null {
  const match = uri.match(/files\/([^:/?#]+)/);
  return match ? match[1] : null;
}

/**
 * Waits until the file reports ACTIVE. Omni puts larger videos in the Files
 * API, and they are not downloadable the instant the call returns.
 */
async function waitForFile(fileId: string, signal: AbortSignal): Promise<void> {
  const key = geminiKey();
  const deadline = Date.now() + 5 * 60_000;

  for (;;) {
    const file = await apiFetch<{ state?: string; error?: { message?: string } }>(
      `${GENAI_BASE}/files/${fileId}?key=${key}`,
      { signal },
    );

    if (file.state === "ACTIVE" || !file.state) return;
    if (file.state === "FAILED") {
      throw new FatalError(
        `Omni: the file failed to process — ${file.error?.message ?? "FAILED"}`,
      );
    }
    if (Date.now() > deadline) {
      throw new RetryableError("Omni: the video took too long to become available");
    }

    await new Promise((resolve) => setTimeout(resolve, 3000));
  }
}

/** Large images make the request fail; 1024px is plenty as a reference. */
async function shrink(data: Buffer): Promise<Buffer> {
  try {
    return await sharp(data)
      .rotate()
      .resize(1024, 1024, { fit: "inside", withoutEnlargement: true })
      .jpeg({ quality: 88 })
      .toBuffer();
  } catch {
    return data;
  }
}

async function omniClip(
  opts: Required<
    Pick<GenerateClipOptions, "prompt" | "aspectRatio" | "durationSeconds">
  > & {
    image?: { data: Buffer; mimeType: string };
  },
  signal: AbortSignal,
): Promise<GeneratedClip> {
  const key = geminiKey();
  if (!key) throw new FatalError("GEMINI_API_KEY is not set");

  const model = omniModel();

  // Omni has no duration parameter — it reads the length from the prompt.
  const promptText = [
    opts.prompt.trim(),
    "",
    `Duration: about ${opts.durationSeconds} seconds. One single continuous shot, no cuts.`,
    "Keep every product, garment, colour, print and face exactly as in the reference image — animate it, do not redesign it.",
    "No on-screen text, no captions, no subtitles, no logos, no watermarks.",
  ].join("\n");

  const input: InteractionContent[] = [];
  if (opts.image) {
    const prepared = await shrink(opts.image.data);
    input.push({
      type: "image",
      mime_type: "image/jpeg",
      data: prepared.toString("base64"),
    });
  }
  input.push({ type: "text", text: promptText });

  const json = await apiFetch<InteractionResponse>(
    `${GENAI_BASE}/interactions?key=${key}`,
    {
      method: "POST",
      signal,
      headers: { "content-type": "application/json" },
      body: JSON.stringify({
        model,
        // With text alone Omni also accepts a plain string, but the array form
        // always works — so there is only ever one code path here.
        input,
        response_format: {
          type: "video",
          aspect_ratio: opts.aspectRatio,
          // Videos are usually larger than 4MB, and inline base64 stops coming
          // back at that size, so always ask for a URI.
          delivery: "uri",
        },
      }),
    },
  );

  if (json.error) {
    throw new FatalError(`Omni: ${json.error.message ?? json.error.status}`);
  }
  if (json.status && json.status !== "completed") {
    throw new RetryableError(`Omni: status ${json.status}`);
  }

  const video = findVideo(json);
  if (!video) throw new Error("Omni returned no video");

  if (video.data) {
    return {
      data: Buffer.from(video.data, "base64"),
      mimeType: video.mime_type ?? "video/mp4",
      provider: `omni:${model}`,
      prompt: opts.prompt,
    };
  }

  const fileId = fileIdFromUri(video.uri ?? "");
  if (!fileId) throw new Error(`Omni returned a video URI we cannot read: ${video.uri}`);

  await waitForFile(fileId, signal);

  const data = await apiFetch<Buffer>(
    `${GENAI_BASE}/files/${fileId}:download?alt=media&key=${key}`,
    { expect: "buffer", signal },
  );

  if (data.length < 10_000) throw new Error("Omni returned an empty file");

  return {
    data,
    mimeType: video.mime_type ?? "video/mp4",
    provider: `omni:${model}`,
    prompt: opts.prompt,
  };
}

/* ------------------------------------------------------------------ *
 *  Public API
 * ------------------------------------------------------------------ */

/**
 * Generates one AI video clip.
 *
 * Throws a ChainError on failure. Callers are expected to catch it and fall
 * back to the still image — a reel must never fail because of this step.
 */
export function generateVideoClip(
  options: GenerateClipOptions,
): Promise<ChainResult<GeneratedClip>> {
  const duration = Math.round(
    Math.min(
      MAX_CLIP_SECONDS,
      Math.max(MIN_CLIP_SECONDS, options.durationSeconds ?? 5),
    ),
  );
  const aspectRatio = options.aspectRatio ?? "9:16";

  return runChain<GeneratedClip>(
    [
      {
        name: "omni",
        label: "Omni — Gemini Omni Flash",
        free: false,
        configured: aiVideoConfigured,
        run: (signal) =>
          omniClip(
            {
              prompt: options.prompt,
              image: options.image,
              aspectRatio,
              durationSeconds: duration,
            },
            signal,
          ),
        // A 10 second clip takes one to two minutes; a heavier render longer.
        timeoutMs: 6 * 60_000,
      },
    ],
    {
      label: "AI video",
      prefer: options.prefer ?? process.env.VIDEO_PROVIDER,
      preferFree: false,
      retries: 1,
      backoffMs: 5000,
    },
  );
}

/** Whether AI video is ready — shown on the Setup page. */
export function aiVideoStatus() {
  const configured = aiVideoConfigured();
  return [
    {
      key: "omni",
      label: `Omni — Gemini Omni Flash (${omniModel()})`,
      free: false,
      configured,
      note: configured
        ? `Up to ${aiVideoMaxClips()} clips per reel — change AI_VIDEO_MAX_CLIPS to adjust.`
        : "Needs GEMINI_API_KEY with billing enabled, since Omni is on the paid tier. Reels still render without it — ffmpeg animates the stills instead.",
    },
  ];
}
